### ISSUE TITLE
Default chunksize with unlimited dimensions leads to huge output files

### ISSUE THREAD

**Issue by dionhaefner** [h5py/h5py#2029]
Over at h5netcdf, we noticed that writing files with unlimited dimensions resulted in ~100x larger file sizes (2GB vs. 20MB) compared to using netCDF4 (https://github.com/h5netcdf/h5netcdf/issues/52). This is caused by the different chunking heuristics used by h5py vs. netCDF4. The netCDF4 one is here:

https://github.com/Unidata/netcdf-c/blob/a57101d4b7dcabf7d477c08eee9a5a126732f702/libhdf5/hdf5var.c#L104-L225

Besides targeting larger chunk sizes than `h5py` (16MB vs. 500kB, with a default cache size of 32MB) it also gives unlimited dimensions a chunk size of 1. `h5py` uses a base size of 1024 which is then halved iteratively (along with all other chunk sizes) until the total size lies within the target range:

https://github.com/h5py/h5py/blob/1d569e69fb8d2b1abea196792bb1f8c948180534/h5py/_hl/filters.py#L348-L349

Now the problem with this is that in all cases I've seen, unlimited dimensions are used for something like a time axis that you append to in every iteration of a simulation. So in my experience a chunk size of 1 makes a lot more sense if you have only 1 unlimited dimension and several fixed dimensions.

We found that in some quick and dirty benchmarks, the h5py heuristic leads to ~10% faster reads and writes than the netCDF4 one *in the absence of unlimited dimensions*, so it seems to work well overall.

To circumvent the issue I would suggest a heuristic similar to this:

```python
def guess_chunk(shape, maxshape, typesize):
    # this part is similar to the netCDF4 heuristic
    num_dim = len(shape)
    num_unlimited = sum(x == 0 for x in shape)
    
    if num_dim == num_unlimited:
        # all dimensions are unlimited, keep current behavior
        unlimited_base_size = 1024
    else:
        unlimited_base_size = 1
    
    shape = tuple((x if x!=0 else unlimited_base_size) for i, x in enumerate(shape))
    # proceed with h5py heuristic
    ...
```

Do you have any experience with this issue, and / or would you be interested in adopting a modified heuristic like this?

**Comment by tacaswell** [comment]
If you use a compression filter does the problem go away?

From the comments in h5netcdf/h5netcdf#52 I infer that you have many relatively short (in the unlimited direction) data sets?

**Comment by kmuehlbauer** [comment]
Invoking compression with `compression="gzip", compression_opts=3, shuffle=True` which should be complementary to the netcdf4 standard `zlib=True` (using complevel=3 and shuffle=True) I get for the example data from https://github.com/h5netcdf/h5netcdf/issues/52

h5py -> 24 MB
netcdf4 -> 16 MB 

This is not 100 times larger but still 50%. The automatic chunk sizes do not fit for the dataset with the unlimited dimension.

**Comment by kmuehlbauer** [comment]
> From the comments in [h5netcdf/h5netcdf#52](https://github.com/h5netcdf/h5netcdf/issues/52) I infer that you have many relatively short (in the unlimited direction) data sets?

Yes. that is correct. The dataset is extended over the unlimited dimension in single steps of size 1. So the best chunk size for this unlimited dimension would be 1.

**Comment by dionhaefner** [comment]
Yes, and my point was more general: unlimited dimensions are typically used for "timestep", where you almost always append 1 at a time.

**Comment by ajelenak** [comment]
Chunks of shape `(time=1, ...)` often perform poorly for those who want to extract time series. I think those who want to use chunking should explicitly specify chunk shape and not rely on any formula for guessing.

**Comment by dionhaefner** [comment]
Chunking is mandatory when using unlimited dimensions, so IMO we should try to supply a sane default.

The current default writes 1024x larger files than needed (in the worst case), which to me sounds worse than degraded reading performance. But then again you probably know better how people use these features in the wild, so if you disagree we can close this.

**Comment by takluyver** [comment]
Going to the other extreme, if you make a dataset with `maxshape=(5, None)`, the suggested heuristic would give it chunks of `(5, 1)`, which is pathologically small, making it slower to read data. Maybe that's not so common, though I've seen EuXFEL datasets with `maxshape=(1, None)` (not created through Python, so they wouldn't be directly affected).

Maybe rather than going straight to 1 for unlimited dimensions, the heuristic should prioritise cutting the unlimited dimensions down, so it would get to `(250, 250, 40, 1)` before trying to cut down the fixed dimensions. That would allow it to still produce a larger chunk if the other dimensions are very small. OTOH, heuristics are hard, and maybe that breaks something else I haven't thought of. It obviously assumes that you'll tend to read & write pieces across (not along) the unlimited dimension.

**Comment by dionhaefner** [comment]
Yes, I agree. It also occurred to me that this would imply radically different chunks for `maxshape=(None,)` and `maxshape=(None, 1)`, which feels icky.

> Maybe rather than going straight to 1 for unlimited dimensions, the heuristic should prioritise cutting the unlimited dimensions down

Excellent suggestion IMO :) Much better than what I came up with.

**Comment by tacaswell** [comment]
I second everything @takluyver said.

> Maybe rather than going straight to 1 for unlimited dimensions, the heuristic should prioritise cutting the unlimited dimensions down

That makes a lot of sense to me, the set of fixed dimensions likely represents some notional "quanta" of data from the users point of view. In turn that strongly suggests that they are going to want to both read and write in multiples of that size so we should try to keep that into a single chunk when possible!

**Comment by dionhaefner** [comment]
Looks like we're on the same page. I could put that into a PR soon unless someone objects.

**Comment by tpongo-afk** [comment]
I'm not sure if that's the problem actually.
One mistake I noticed is that the definition of unlimited dimensions is bad.
Line 349 in filters.py should be:
`shape = tuple((x if y is not None else 1024) for x, y in enumerate(shape, maxshape))`
That is, a dimension is unlimited when the maxshape is None, not when the shape is 0 along the axis.
Supposing that maxshape is a tuple, although it can still be True in principle because of the following part: (the True possibility is not documented)
https://github.com/h5py/h5py/blob/b3697ab2ced3342b3462fc3dcd1a11f26af74f1c/h5py/_hl/filters.py#L247-L253

**Comment by ChrisBarker-NOAA** [comment]
Not sure what the status of this is -- but a bit of history:

Way back when netcdf4 was new, I discovered that he default chunking for 1-D variables was horrendously bad.

It turns out that it was always setting the chunk size in the first (in this case only) dimension to 1 (or maybe it was the unlimited dimension, same thing it this case) -- so for 1-D variables you got 1-element chunks -- really really bad! 

Anyway, I can't remember exactly the solution, but something along the line of making sure the default chunks were always at least some minimum size was very helpful.

I'm always unsure of what defaults should be -- I *think* at a glance, that @dionhaefner 's suggested heuristic would make  more or less "square" chunks -- which may be the best option for when you don't know the expected access patterns.

Though making the first dimension (usually time) small (e.g. chunk size 1) is consistent with (old) folks' expectation that arrays are in C order :-)

One Note:
>  the set of fixed dimensions likely represents some notional "quanta" of data from the users point of view.

I don't think so at all -- fixed dimensions are for when the size is known ahead of time, which is an independent concept. For example, I mostly use results from gridded oceanographic models -- we know the grid size up front, so those dimensions are fixed -- but that doesn't mean the user wants the entire grid at once -- the the example of a time series at a point.

Unlimited dimensions are often used, as discussed, for when you are adding the data bit by bit -- e.g. time. So you'd think a chunk of one on unlimited dimensions would make sense -- except that reading patterns may be completely different than writing patterns -- again the time-series-at-a-point example.

The fact is there is no "best" chunking scheme for arbitrary use patterns.

**Comment by tacaswell** [comment]
> I don't think so at all --

That is fair, I mostly think about 2D detectors where the framesize is very much the quanta, but for huge grids that is clearly not the case!

**Comment by ChrisBarker-NOAA** [comment]
Hmm -- I found this issue 'cause I'm looking at problems with the default chunk cache size in netCDF4. But anyway, it's making me think about the nightmare of chunking again!

But a new idea. As mentioned -- there is no default that is good for all use cases, period.

So currently, users are left to just use the defaults and hope for the best (many many people do this) -- or figure out what they need, and setting the chunking by hand -- but that requires a fair bit of understanding of the issue.

Perhaps libraries (like h5py, netCDF4, zarr? [*]) could a provide a set of purpose-built defaults -- so you could simply say:

set_chunks_to('time_series_analysis')

The question is -- is there a small and well defined enough set of access patterns that would make sense?

[*] Now that I'm thinking about it -- this might be a good candidate for a "chunking" lib that multiple projects would cooperate on.

**Comment by takluyver** [comment]
> a provide a set of purpose-built defaults -- so you could simply say:
>
> `set_chunks_to('time_series_analysis')`
>
> The question is -- is there a small and well defined enough set of access patterns that would make sense?

My instinct is that a small set of simple access patterns probably covers, say, 95% of cases, but that there won't be clear, sensible, unambiguous names to describe these.

I could imagine providing a way to indicate that the chunks should preferentially cut along *these* axes - e.g. the detector frames axis in the photon science examples familiar to @tacaswell & myself. Or that chunks should prefer not to cut up *those* axes - e.g. detector pixel axes. Something like:

```python
f.create_dataset('ds', (frames, px1, px2), chunks_split_axes=(0,))
f.create_dataset('ds', (frames, px1, px2), chunks_keep_axes=(1, 2))
```

But this still needs the person making the file to be aware of the issue and think about how chunks should work. And I don't think it's any clearer, either to write or to read, than an explicit chunk shape:

```python
f.create_dataset('ds', (frames, px1, px2), chunks=(1, px1, px2))
```

We could give up on guessing entirely ("In the face of ambiguity...") and force people creating chunked datasets to specify the chunk shape explicitly. So it would be an error to e.g. enable data compression without passing a chunk shape. Or we could split the difference and allow something like `chunks='auto'` to opt in to the default guessing. But IDK if it's worth the loss of convenience for cases where the default guess is good enough.
